### PR TITLE
Ditch item grid

### PR DESCRIPTION
**Pull request by BTMorton**

### CODE REVIEW

File: src/directives/NgGrid.ts [hunk 1]
```diff
@@ -902,108 +941,103 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
...
 
-						if (this._itemGrid[row][col] != null) {
-							const item: NgGridItem = this._itemGrid[row][col];
-							if (item.isFixed) continue;
+					let lowestRowForItem: number = lowestRowPerColumn.has(itemPos.col) ? lowestRowPerColumn.get(itemPos.col) : 1;
 
-							const itemDims: NgGridItemSize = item.getSize();
-							const itemPos: NgGridItemPosition = item.getGridPosition();
+					for (let i: number = 1; i < itemDims.x; i++) {
+						const lowestRowForColumn = lowestRowPerColumn.has(itemPos.col + i) ? lowestRowPerColumn.get(itemPos.col + i) : 1;
```
Comment (aebsubis) [inline src/directives/NgGrid.ts]: I thinik that "const lowestRowForColumn = lowestRowPerColumn.get(itemPos.col + i) || 1;" can be used too.

File: src/directives/NgGrid.ts [hunk 2]
```diff
@@ -1016,28 +1050,35 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
...
-					case 'up':
-					case 'down':
-					default:
-						pos.row = Math.max.apply(null, collisions.map((item: NgGridItem) => item.row + item.sizey));
-						break;
-					case 'left':
-					case 'right':
-						pos.col = Math.max.apply(null, collisions.map((item: NgGridItem) => item.col + item.sizex));
-						break;
+				let fixDirection: NgConfigFixDirection = this._itemFixDirection;
```
Comment (aebsubis) [inline src/directives/NgGrid.ts]: The method mentioned earlier could be reused here to calculate the computed fix direction.

File: src/directives/NgGrid.ts [hunk 3]
```diff
@@ -1103,38 +1144,13 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
...
-
-				if (this._items.indexOf(this._itemGrid[y][x]) < 0 || !withinRow || !withinCol) {
-					delete this._itemGrid[y][x];
-				}
-			}
-
-			if (Object.keys(this._itemGrid[y]).length == 0) {
-				delete this._itemGrid[y];
-			}
+		if (this._itemsInGrid.has(item.uid)) {
```
Comment (aebsubis) [inline src/directives/NgGrid.ts]: I think we can delete directly.

File: src/directives/NgGrid.ts [hunk 4]
```diff
@@ -1215,17 +1231,20 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
 	}
 
 	private _getItemFromPosition(position: NgGridRawPosition): NgGridItem {
-		for (let item of this._items) {
+		let matchedItem: NgGridItem = null;
+		this._itemsInGrid.forEach((item) => {
```
Comment (aebsubis) [inline src/directives/NgGrid.ts]: Why not use .find?

File: src/directives/NgGrid.ts [hunk 5]
```diff
@@ -1013,36 +1044,91 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
...
 
-			if (!this._isWithinBoundsY(pos, dims)) {
-				pos.col++;
+				pos.col = Math.max(pos.col + 1, Math.min.apply(Math, itemsInPath.map((item) => item.col + dims.x)));
 				pos.row = 1;
 			}
-			if (!this._isWithinBoundsX(pos, dims)) {
-				pos.row++;
+		} else if (this._itemFixDirection === "horizontal") {
+			fixLoop:
```
Comment (aebsubis) [inline src/directives/NgGrid.ts]: I'm not a fan of this, but if it gets the job done I can't complain :smile:

Comment (BTMorton) [reply src/directives/NgGrid.ts]: What would you propose instead?

**Pull request merged**